### ISSUE TITLE
New release?

### ISSUE THREAD

**Issue by NeKzor** [denosaurs/byte_type#45]
It's been a while and the new sized types look cool. Although I think the [calculateTotalSize](https://github.com/denosaurs/byte_type/blob/541df5ba28d282dfcef45ad5441d19cf9b391bbc/src/util.ts#L25C14-L25C32) util should be called `sizeOf` because it's a common term in C like languages. Thoughts? Also a util like `offsetOf` would be useful so I can remove my current hack.

**Comment by eliassjogreen** [comment]
I agree about the renaming of those utils. I could do it some evening when I got time this week :)

**Comment by MierenManz** [comment]
@eliassjogreen Should I create a new release? I want to land one more pr real quick. It's work for `Viewable<T>` and `asView` but also a neat feature that was easy to add (it does still need optimization)

**Comment by MierenManz** [comment]
@NeKzor What is your current hack and what should `offsetOf` do? I assume it's something like this?
```ts
const codec = new SizedStruct({ byte: u8, word: u16 });

console.log(offsetOf(codec)) // Logs `{ aligned: 4, packed: 3 }`
```

Edit: the idea to rename `getTotalSize` to `sizeOf` is a good one! I like it
Edit 2: We could probably rename `getBiggestAlignment` to `alignmentOf` aswell. That seems to convey the message aswell.

**Comment by eliassjogreen** [comment]
@MierenManz Sure, let's make a new release after #49 has landed :)

**Comment by NeKzor** [comment]
I better not post my hack because I implemented it incorrectly lol

It would get the offset of a field inside a struct inspired by [C++](https://en.cppreference.com/w/cpp/types/offsetof) and [Rust](https://doc.rust-lang.org/nightly/core/mem/macro.offset_of.html). Something like this:
```ts
offsetOf({ byte: u8, word: u32le }, 'word'); // 4
offsetOfPacked({ byte: u8, word: u32le }, 'word'); // 1
```

Also I have another idea for declaring sized arrays. How about mapping a tuple to `SizedArrayType`/`ArrayType` or maybe even `TypedArray`?
Here is an example:
```ts
new Struct({
    foo: [u8, 100], // length 100
});
```

**Comment by MierenManz** [comment]
> Also I have another idea for declaring sized arrays. How about mapping a tuple to SizedArrayType/ArrayType or maybe even TypedArray?
Here is an example:
```ts
new Struct({
    foo: [u8, 100], // length 100
});
```
I've considered stuff like this in the past and eventho I really like how much cleaner it is compared to `new SizedArrayType(u8, 100)`. I am conflicted if this magic is something we should want for the library.

**Comment by MierenManz** [comment]
> It would get the offset of a field inside a struct inspired by [C++](https://en.cppreference.com/w/cpp/types/offsetof) and [Rust](https://doc.rust-lang.org/nightly/core/mem/macro.offset_of.html). Something like this:
```ts
offsetOf({ byte: u8, word: u32le }, 'word'); // 4
offsetOfPacked({ byte: u8, word: u32le }, 'word'); // 1
```
How would this work for any (sized) type that encodes/decodes a primitive? I assume the second field would just be `undefined` or `null`?

**Comment by MierenManz** [comment]
> > Also I have another idea for declaring sized arrays. How about mapping a tuple to SizedArrayType/ArrayType or maybe even TypedArray?
> > Here is an example:
> 
> new Struct({
>     foo: [u8, 100], // length 100
> });
> 
> I've considered stuff like this in the past and eventho I really like how much cleaner it is compared to `new SizedArrayType(u8, 100)`. I am conflicted if this magic is something we should want for the library.

Maybe this is something we could do with decorators so that the magic spaghetti for it would be in a optional decorator somehow?

**Comment by MierenManz** [comment]
Closing because 0.4.1 got released.

I have created #50 and #51 so that we can continue effective discussion about `offsetOf` utility functions and the sugarcoat constructing of compound types.
